Add more preload functionality

TPGamesNL opened the pull request:
### Description
This PR adds the `PostStructureCreateEvent`, which is very similar to the `PreScriptLoadEvent`, but the `PostStructureCreateEvent` is called after the structure of a script is created, which happens before any of the scripts are loaded. 
With this PR, addon developers can add their own preload functionality, similar to how function signatures are created before any of the scripts are loaded. 

Quick test plugin: [TestPlugin-1.0-SNAPSHOT.jar](https://gofile.io/d/ktmsUJ) ([source](https://gist.github.com/TPGamesNL/2d3c9167fa6a11ed8a979290a8cc8b12))
Output: 
![image](https://user-images.githubusercontent.com/29547183/85230831-26f16300-b3f3-11ea-98f3-cdb1f272dd6a.png)

---
**Target Minecraft Versions:** Any
**Requirements:** None
**Related Issues:** None

Pikachu920 commented:
I can't really read the dif because of the line endings @JRoy mentioned so I may be missing it but I do wonder how this solves the issue describe in https://github.com/TPGamesNL/skript-mirror/issues/14. wouldn't it make more sense to just make `PreScriptLoadEvent` provide a list of configs rather than a single config at a time when possible?

JRoy commented on src/main/java/ch/njol/skript/ScriptLoader.java:
```diff
@@ -939,6 +984,9 @@ public String run(final Matcher m) {
 					}
 				}
 				
+				if (callPostStructureEvent)
+					Bukkit.getPluginManager().callEvent(new PostStructureCreateEvent(config));
+				
```
Use brackets for codestyle

JRoy commented on src/main/java/ch/njol/skript/ScriptLoader.java:
```diff
@@ -158,6 +159,50 @@ public static boolean getCallPreloadEvent() {
+	public static void setCallPostStructureEvent(boolean state, SkriptAddon addon) {
+		Validate.notNull(addon);
+		callPostStructureEvent = state;
+		if (state)
+			postStructureListeners.add(addon);
+		else
+			postStructureListeners.remove(addon);
```
brackets

JRoy commented on src/main/java/ch/njol/skript/ScriptLoader.java:
```diff
@@ -158,6 +159,50 @@ public static boolean getCallPreloadEvent() {
+	 * but before the script is parsed.
+	 */
+	private static boolean callPostStructureEvent;
+
+	/**
+	 * A set of all the SkriptAddons that have called
+	 * {@link ScriptLoader#setCallPostStructureEvent(boolean, SkriptAddon)}
+	 * with true.
+	 */
+	private static Set<SkriptAddon> postStructureListeners = new HashSet<>();
+
+	/**
+	 * Sets {@link ScriptLoader#callPostStructureEvent} to the provided boolean,
+	 * and adds/removes the provided SkriptAddon from {@link ScriptLoader#postStructureListeners}
+	 * depending on the provided boolean (true adds, false removes).
+	 * @param state The new value for {@link ScriptLoader#postStructureListeners}
+	 * @param addon A non-null SkriptAddon
+	 */
+	public static void setCallPostStructureEvent(boolean state, SkriptAddon addon) {
+		Validate.notNull(addon);
+		callPostStructureEvent = state;
+		if (state)
+			postStructureListeners.add(addon);
+		else
+			postStructureListeners.remove(addon);
+	}
+
+	public static boolean getCallPostStructureEvent() {
+		return callPostStructureEvent;
+	}
+
+	/**
+	 * Returns an unmodifiable list of all the addons
+	 * that have called {@link ScriptLoader#setCallPostStructureEvent(boolean, SkriptAddon)}
+	 * with true.
+	 */
+	@SuppressWarnings("null")
+	public static Set<SkriptAddon> getPostStructureListeners() {
+		return Collections.unmodifiableSet(postStructureListeners);
+	}
```
I don't understand the use case for this here, why not always call the event. If nobody is listening to it, there's very little overhead anyway (especially on Paper). The Set of addons seems kind of pointless since they're really not useful.

APickledWalrus replied:
I agree with this. I was looking at this last night (we talked about it in discord a little), and this seemed kind of pointless. It also seems like an addon could disable the event being called, which could break other addons that depend on it.

Pikachu920 replied:
it's to be consistent with how i implemented the preload event and i implemented the preload event that way because there is a similar toggle to not call function events. looking back though i wouldn't have done this now

TPGamesNL commented:
Finally fixed the line endings, sorry for the pile of commits. Tell me if it'd be better to remake the PR, to not have 16 commits for 2 files.


> I can't really read the dif because of the line endings @JRoy mentioned so I may be missing it but I do wonder how this solves the issue describe in [TPGamesNL/skript-mirror#14](https://github.com/TPGamesNL/skript-mirror/issues/14). wouldn't it make more sense to just make `PreScriptLoadEvent` provide a list of configs rather than a single config at a time when possible?

Besides the fact that it'd be a breaking change for addons using it, I think it would also require the code in ScriptLoader to be changed a bit, since the structure loading isn't very centralized, it seems to be distributed over 3 different methods (`loadStructures(File[])`, `loadStructures(File)` and `loadStructure(File directory)`), which are (some, not all) recursive and they call eachother. Also, the issue described in https://github.com/TPGamesNL/skript-mirror/issues/14 is a little vague, but I generally mean the same as pre-loading function signatures, so that you don't receive errors about non-existent functions if you've declared them later then using them in the code.
Another possibility would be changing the `PreScriptLoadEvent` so that it gets called after the structure generating (where `PostStructureCreateEvent` is now called), but that could possibly have side effects for addons using it.

> Use brackets for codestyle

> brackets

> I don't understand the use case for this here, why not always call the event. If nobody is listening to it, there's very little overhead anyway (especially on Paper). The Set of addons seems kind of pointless since they're really not useful.

> I agree with this. I was looking at this last night (we talked about it in discord a little), and this seemed kind of pointless. It also seems like an addon could disable the event being called, which could break other addons that depend on it.

I agree. The reason I added it, was because of the similarity with PreScriptLoadEvent, which uses this exact system + code style (I copied all of it, and changed the names) (PreScriptLoadEvent + its registration added [here](https://github.com/SkriptLang/Skript/pull/1150)). I'm fine with removing the registration stuff, as I also think it has very little use, including for the PreScriptLoadEvent.
I'm fine with removing the whole registration stuff, but does that mean I should also remove it for the `PreScriptLoadEvent`?

Pikachu920 commented:
>Besides the fact that it'd be a breaking change for addons using it, I think it would also require the code in ScriptLoader to be changed a bit, since the structure loading isn't very centralized, it seems to be distributed over 3 different methods (loadStructures(File[]), loadStructures(File) and loadStructure(File directory)), 

yeah agreed on the breaking change however I don't think there are any addons actually using this at the moment (plus it could be implemented in a backwards compatible way if it is really needed in the end. as far as the `loadStructures` and `loadScripts` methods, I believe they are all overloads which eventually call `loadScript(Config)`. right?

TPGamesNL commented:
> yeah agreed on the breaking change however I don't think there are any addons actually using this at the moment (plus it could be implemented in a backwards compatible way if it is really needed in the end. as far as the loadStructures and loadScripts methods, I believe they are all overloads which eventually call loadScript(Config). right?

I agree that there are probably not many addons using it.
About the `loadScript(Config)` however, yes, every loading code points to that method, but that defeats the whole purpose of this PR, as this is the way `PreScriptLoadEvent` is currently implemented. The reason behind this PR was that the 
`PreScriptLoadEvent` for `b.sk` was only called after `a.sk` was fully loaded. Because of that, any content from `b.sk` isn't able to affect the loading for `a.sk`, which is what I was aiming for.

Pikachu920 commented:
but you could make it so that the event is called for both a.sk and b.sk at the same time

TPGamesNL commented:
I made it work with calling the event in `loadScripts(List<Config> configs)`. Should I keep it backwards compatible or should I remove all the registration?

TPGamesNL commented:
I removed the new event, and instead removed all the registration nonsense, changed where `PreScriptLoadEvent` is called and made it use a list of configs instead of a single one

TPGamesNL commented:
> def a better solution, but currently the event won't be called if `ScriptLoader#loadScript` is used directly to load a single script

True, but `loadScripts` has the only usage of the method, and `loadScript` is private, so shouldn't be called from addons. I'll at a comment for future edits to call the event if the method is called.

ShaneBeee commented:
Merged! Thank you for your contribution!!!
